v1 APIs and Stable Profile announcement (#14981)
istio/istio.io: content/en/blog/2024/v1-apis/index.md

| | [Peer Authentication](/docs/reference/config/security/peer_authentication/) |  `v1`, `v1beta1` |
| | [Request Authentication](/docs/reference/config/security/request_authentication/) |  `v1`, `v1beta1` |
| Telemetry | [Telemetry](/docs/reference/config/telemetry/) |  `v1`, `v1alpha1` |
| Extension | [Wasm Plugin](/docs/reference/config/proxy_extensions/wasm-plugin/) |  `v1alpha1` |

Istio can also be configured [using the Kubernetes Gateway API](/docs/setup/additional-setup/getting-started/).

## Using the `v1` Istio APIs

There are some APIs in Istio that are still under active development and are subject to potential changes between releases. For instance, the Envoy Filter, Proxy Config and Wasm Plugin APIs.

Furthermore, Istio maintains a strictly identical schema across all versions of an API due to limitations in [CRD versioning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/). Therefore, even though there is a `v1` Telemetry API, the three `v1alpha1` fields mentioned [above](#telemetry-api) can still be utilized when declaring a `v1` Telemetry API resource.

> For risk-averse environments, we have provided a sample `Stable Validation Policy` that can ensure only `v1` APIs and fields are used.
> 
> In new environments, selecting the `Stable Validation Policy` upon installing Istio will guarantee that all future Custom Resources created or updated are `v1` and contain only `v1` features.

Reviewer: ```suggestion
For risk-averse environments, we have added a **stable validation policy**, a [validating admission policy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) which can ensures that only `v1` APIs and fields are used with Istio APIs.

In new environments, selecting the stable validation policy upon installing Istio will guarantee that all future Custom Resources created or updated are `v1` and contain only `v1` features.
```

LGTM overall!  Just a couple of fixes that were missed due to them being attached to questions and this is good to go from my end.

Thanks very much for writing this Whitney!

---

> title: "Getting Started with using v1 Istio APIs "

Reviewer 2: Bit of a nit, but if we're considering [pushing new users toward Gateway API](https://github.com/istio/istio.io/issues/14975) and Ambient, maybe we should use a phrase other than "getting started" here? Not a strong opinion though, could also be fine to leave as-is.

Author: Updated to Introducing Istio v1 APIs

* `metrics.reportingInterval`
    * Reporting interval allows configuration of the time between calls out to for metrics reporting. This currently only supports TCP metrics but we may use this for long duration HTTP streams in the future.
    * At this time, Istio lacks usage data to support the need for this feature.
* `accessLogging.filter`
    * If specified, this filter will be used to select specific requests/connections for logging.
    * This feature is based on a relatively new feature in Envoy, and Istio needs to further develop the use case and implementation before graduating it to `v1`.
* `tracing.useRequestIdForTraceSampling`
    * This value is true by default. The format of this Request ID is specific to Envoy, and if the Request ID generated by the proxy that receives user traffic first is not specific to Envoy, Envoy will break the trace because it cannot interpret the Request ID. By setting this value to false, we can prevent [Envoy from sampling based on the Request ID](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/observability/tracing#trace-context-propagation).

>     * There is not a strong use case for making this configurable through the Telemetry API.

Reviewer 3: It would be great to have a couple of sentences here telling users to provide feedback via GitHub if they feel strongly about this fields

Author: Added!

> When migrating Telemetry Resources to the `v1` Telemetry API, we strongly recommend removing the usage of the non-`v1` fields. Not doing this will not cause any disruptions, as Istio has chosen to maintain the same schema across API Versions to avoid unyielding API conversion issues. This is a known, complex issue with [Kubernetes API Versioning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#overview), and the Gateway API has chosen to address this with [Release Channels](https://gateway-api.sigs.k8s.io/concepts/versioning/).

Reviewer 3: "Not doing this will not cause any disruptions" - this is a double negative and a bit confusing. What are you advising users to do here and for what reason?

Reviewer 2: ```suggestion
```

I think we could actually just cut this entire paragraph - the more user-focused instructions below under "Using the v1 Istio APIs" feel sufficient for managing practical concerns, and getting into the "why" and "how" here feels a bit distracting for a primarily end-user audience.

Author: Nixed!

| Istio [Gateway](/docs/reference/config/networking/gateway/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Proxy Config](/docs/reference/config/networking/proxy-config/) |  `v1beta1` |
| [Service Entry](/docs/reference/config/networking/service-entry/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Sidecar](/docs/reference/config/networking/sidecar/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Virtual Service](/docs/reference/config/networking/virtual-service/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Workload Entry](/docs/reference/config/networking/workload-entry/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Workload Group](/docs/reference/config/networking/workload-group/) |  `v1`, `v1beta1`, `v1alpha3` |
| **Extension APIs** |
| [Wasm Plugin](/docs/reference/config/proxy_extensions/wasm-plugin/) |  `v1alpha1` |

## Using the `v1` Istio APIs

As seen above, there are a few non-`v1` APIs and fields. The Envoy Filter, Proxy Config and Wasm Plugin aren’t `v1`, and within the Telemetry API, there are 3 non-`v1` fields.

Since Istio maintains the same schema across API versions, users can utilize non-`v1` fields in their `v1` Custom Resources.

> For risk-averse environments, we have provided a sample `Stable Profile` that can ensure only `v1` APIs and fields are used.

Reviewer 3: We should expand on this a bit (and maybe even change the heading to reflect the release channels language @craigbox was suggesting). We should talk about CEL and Kubernetes versioning, and _why_ we decided to create an experimental ValidatingAdmissionPolicy (and say that this feature is only compatible with kubernetes 1.30).

Reviewer 2: I actually kinda like the current more direct framing focused on "how do I use this" without getting into too much complex background - the "why" is interesting, but I don't think hugely relevant to most Istio end-users.

Adding a sentence addressing how platform teams can enforce or extend these policies for their app dev team customers would be useful though, and agreed we need to mention Kubernetes 1.30+ (or 1.26+ with the `ValidatingAdmissionPolicy` feature gate enabled?)

We should be careful to avoid confusion with Istio's [configuration profiles](https://istio.io/latest/docs/setup/additional-setup/config-profiles/) though, I'd suggest avoiding the term "profile" here.

```suggestion
For risk-averse environments, we have provided a sample `Stable Admission Policy` that can ensure only `v1` APIs and fields are used.
```

Reviewer 3: There is a new Istio profile though technically. I agree we probably shouldn't advertise it though

---
title: "Introducing Istio v1 APIs "
description: Announcing the v1 Networking, Security and Telemetry APIs. 
publishdate: 2024-05-13
attribution: Whitney Griffith - Microsoft
keywords: [istio, traffic, security, telemetry, API]
target_release: 1.22
---

> Istio has several [Networking](/docs/reference/config/networking/), [Security](/docs/reference/config/security/) and [Telemetry](/docs/reference/config/telemetry/) APIs that are crucial for ensuring the secure, seamless connectivity, and effective monitoring of services within the Service Mesh layer. These APIs are used on thousands of clusters across the world, securing and enhancing critical infrastructure. As a reflection of the stability, adoption, and value of these resources, the Istio community has decided to promote many of our CRDs to `v1` in Istio 1.22.

Reviewer 3: nit: consistent parallelism and capitalization
```suggestion
Istio has several [Networking](/docs/reference/config/networking/), [Security](/docs/reference/config/security/) and [Telemetry](/docs/reference/config/telemetry/) APIs that are crucial for ensuring the robust security, seamless connectivity, and effective monitoring of services within the service mesh layer. These APIs are used on thousands of clusters across the world, securing and enhancing critical infrastructure. As a reflection of the stability, adoption, and value of these resources, the Istio community has decided to promote many of our CRDs to `v1` in Istio 1.22.
```

In Istio 1.22 we are happy to announce that a concerted effort has been made to graduate the below APIs to `v1`:
* [Destination Rule](/docs/reference/config/networking/destination-rule/)
* [Gateway](/docs/reference/config/networking/gateway/)
* [Service Entry](/docs/reference/config/networking/service-entry/)
* [Sidecar](/docs/reference/config/networking/sidecar/)
* [Virtual Service](/docs/reference/config/networking/virtual-service/)
* [Workload Entry](/docs/reference/config/networking/workload-entry/)
* [Workload Group](/docs/reference/config/networking/workload-group/)
* [Telemetry API](/docs/reference/config/telemetry/)*
* [Peer Authentication](/docs/reference/config/security/peer_authentication/)

## Telemetry API

> The `v1` Telemetry API is the only API that was promoted that had changes from its previous version. The following features weren’t promoted to `v1`:

Reviewer 4: Worth clarify if below are the changes from previous version thus not being promoted to v1.

Author: @linsun does this help? 

_The `v1` Telemetry API is the only API that was promoted that had changes from its previous version. The following `v1alpha1` features weren’t promoted to `v1`_

Reviewer 4: Thanks, yep that is better. Another clarification - previous version meaning previous API version, not Istio version?

Author: Correct, previous API version! I have updated it.

| [Destination Rule](/docs/reference/config/networking/destination-rule/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Envoy Filter](/docs/reference/config/networking/envoy-filter/) |  `v1alpha3` |
| Istio [Gateway](/docs/reference/config/networking/gateway/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Proxy Config](/docs/reference/config/networking/proxy-config/) |  `v1beta1` |
| [Service Entry](/docs/reference/config/networking/service-entry/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Sidecar](/docs/reference/config/networking/sidecar/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Virtual Service](/docs/reference/config/networking/virtual-service/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Workload Entry](/docs/reference/config/networking/workload-entry/) |  `v1`, `v1beta1`, `v1alpha3` |
| [Workload Group](/docs/reference/config/networking/workload-group/) |  `v1`, `v1beta1`, `v1alpha3` |
| **Extension APIs** |
| [Wasm Plugin](/docs/reference/config/proxy_extensions/wasm-plugin/) |  `v1alpha1` |

## Using the `v1` Istio APIs

As seen above, there are a few non-`v1` APIs and fields. The Envoy Filter, Proxy Config and Wasm Plugin aren’t `v1`, and within the Telemetry API, there are 3 non-`v1` fields.

> Since Istio maintains the same schema across API versions, users can utilize non-`v1` fields in their `v1` Custom Resources.

Reviewer 4: I don't think this applies to Envoy Filter, Proxy Config and Wasm Plugin, so felt odd to have this after it.

Author: I have updated the wording, does it make more sense now?

Reviewer 4: I think it may make sense to cover non-`v1` fields in their `v1` Custom Resources first, then cover non-v1 CRs next which is `EnvoyFilter` etc.

Author: PTAL, I have made changes to give more insights.

Reviewer: LGTM; I understand now that it's not a decision on our part, it's just a side effect of how Kubernetes works.

For risk-averse environments, we have provided a sample `Stable Validation Policy` that can ensure only `v1` APIs and fields are used.

> In new environments, selecting the `Stable Validation Policy` upon installing Istio will guarantee that all future Custom Resources created or updated are `v1` and contain only `v1` features.

Author: Until `v2` becomes a reality, I don't think we need to inflate the definition of the Stable Validation Policy.

> In new environments, selecting the `Stable Validation Policy` upon installing Istio will guarantee that all future Custom Resources created or updated are `v1` and contain only `v1` features.

Reviewer 3: I think I agree with Whitney; with a v2 API, all bets are off w.r.t to compatibility IMO

> In new environments, selecting the `Stable Validation Policy` upon installing Istio will guarantee that all future Custom Resources created or updated are `v1` and contain only `v1` features.

Reviewer: cool - just wasn't sure how the policy worked!

If the `Stable Validation Policy` is deployed into an existing Istio Cluster that has Custom Resources not complying with the `Stable Validation Policy`, the only allowed action is to delete the resource or remove the usage of the offending fields.

To install Istio with the `Stable Validation Policy`use the following command:

{{< text bash >}}
$ helm install istio-base -n istio-system --set experimental.stableValidationPolicy=true
{{< /text >}}

To set a specific revision when installing Istio with the `Stable Validation Policy` use the following command:

{{< text bash >}}
$ helm install istio-base -n istio-system --set experimental.stableValidationPolicy=true -set revision=x
{{< /text >}}

> This feature is only compatible with [Kubernetes 1.30](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/). The validations are created using [CEL](https://github.com/google/cel-spec) expressions. Users can modify the validations in the sample `Stable Validation Policy` for their specific needs.

Reviewer: ```suggestion
This feature is compatible with [Kubernetes 1.30](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) and higher. The validations are created using [CEL](https://github.com/google/cel-spec) expressions, and users can modify the validations for their specific needs.
```

## Summary

> In summary, Istio is committed to delivering stable APIs and features essential for the successful operation of the Istio Service Mesh. We would love to receive your feedback to help guide us in making the right decisions as we continue to refine relevant use cases and stability blockers for our features. Please share your feedback by creating [issues](https://github.com/istio/istio/issues), posting in the relevant [Istio Slack](https://slack.istio.io/) channel, or by joining us in our weekly [Combined Working Group](https://github.com/istio/community/blob/master/WORKING-GROUPS.md#working-group-meetings) meeting.

Reviewer: ```suggestion
The Istio project is committed to delivering stable APIs and features essential for the successful operation of your service mesh. We would love to receive your feedback to help guide us in making the right decisions as we continue to refine relevant use cases and stability blockers for our features. Please share your feedback by creating [issues](https://github.com/istio/istio/issues), posting in the relevant [Istio Slack channel](https://slack.istio.io/), or by joining us in our weekly [working group meeting](https://github.com/istio/community/blob/master/WORKING-GROUPS.md#working-group-meetings).
```
